src/merkle.rs

```diff
@@ -204,210 +401,553 @@ impl<T: Element, A: Algorithm<T>, K: Store<T>, U: Unsigned> MerkleTree<T, A, K,
+        for config in configs {
+            let data = S::new_with_config(
+                get_merkle_tree_len(leafs, branches)?,
+                branches,
+                config.clone(),
+            )
+            .context("failed to create data store")?;
+            trees.push(MerkleTree::<E, A, S, BaseTreeArity>::from_data_store(
+                data, leafs,
+            )?);
+        }
+
+        MerkleTree::from_trees(trees)
+    }
+
+    #[inline]
+    fn build_partial_tree(
+        mut data: VecStore<E>,
+        leafs: usize,
+        height: usize,
+    ) -> Result<MerkleTree<E, A, VecStore<E>, BaseTreeArity>> {
+        let root = VecStore::build::<A, BaseTreeArity>(&mut data, leafs, height, None)?;
+        let branches = BaseTreeArity::to_usize();
+
+        let tree_len = get_merkle_tree_len(leafs, branches)?;
+        ensure!(tree_len == Store::len(&data), "Inconsistent tree data");
 
-        // Sanity check: if the `MerkleTree` lost its integrity and `data` doesn't match the
-        // expected values for `leafs` and `height` this can get ugly.
         ensure!(
-            lemma.len() == get_merkle_proof_lemma_len(self.height, branches),
-            "Invalid proof lemma length"
+            is_merkle_tree_size_valid(leafs, branches),
+            "MerkleTree size is invalid given the arity"
         );
-        ensure!(path.len() == self.height - 1, "Invalid proof path length");
 
-        Proof::new(lemma, path)
+        Ok(MerkleTree {
+            data: Data::BaseTree(data),
+            leafs,
+            len: tree_len,
+            height,
+            root,
+            _a: PhantomData,
+            _e: PhantomData,
+            _bta: PhantomData,
+            _sta: PhantomData,
+            _tta: PhantomData,
+        })
     }
 
-    /// Generate merkle tree inclusion proof for leaf `i` by first
-    /// building a partial tree (returned) along with the proof.
-    /// Return value is a Result tuple of the proof and the partial
-    /// tree that was constructed.
-    #[allow(clippy::type_complexity)]
-    pub fn gen_proof_and_partial_tree(
+    /// Generate merkle sub tree inclusion proof for leaf `i` for
+    /// either the top layer or the sub-tree layer, specified by the
+    /// top_layer flag
+    #[inline]
+    fn gen_sub_tree_proof(
         &self,
         i: usize,
-        levels: usize,
-    ) -> Result<(Proof<T, U>, MerkleTree<T, A, VecStore<T>, U>)> {
-        ensure!(
-            i < self.leafs,
-            "{} is out of bounds (max: {})",
-            i,
-            self.leafs
-        ); // i in [0 .. self.leafs]
+        top_layer: bool,
+        arity: usize,
+    ) -> Result<Proof<E, BaseTreeArity>> {
+        ensure!(arity != 0, "Invalid sub-tree arity");
+
+        // Locate the sub-tree the leaf is contained in.
+        let tree_index = i / (self.leafs / arity);
+
+        // Generate the sub tree proof at this tree level.
+        let sub_tree_proof = if top_layer {
```

Comment:
this could use the enum match as well

Reply:
The match works if it's SubTree/TopTree, but can't possibly provide the right type if it's BaseTree, which also has to be handled. Not sure of how best to do that other than what's there.

Reply:
oh sorry, brain fart on my side..